crates/storage/src/cache.rs

```diff
@@ -1,252 +1,150 @@
-        let mut cache = self.cache.lock();
-        let len = cache.len();
-        cache.clear();
+        let len = self.get_len();
+        let mut cache_guard = self.cache.write();
+        cache_guard.clear();
+
+        // Remove mutexes
+        let mut fetch_guard = self.fetch_mutexes.write();
+        fetch_guard.clear();
+
         len
     }
 
-    /// Inserts an entry into the table, dropping the previous value.
-    #[allow(dead_code)]
-    pub(crate) fn insert(&self, k: K, v: V) {
-        let slot = Arc::new(RwLock::new(SlotState::Ready(v)));
-        self.cache.lock().put(k, slot);
+    pub(crate) fn get_len(&self) -> usize {
+        self.cache.read().len()
     }
 
-    /// Returns a clone of an entry from the cache or possibly invoking some function returning a
-    /// `oneshot` channel that will return the value from the underlying database.
-    ///
-    /// This is meant to be used with the `_chan` functions generated by the db ops macro in the
-    /// `exec` module.
-    // https://github.com/rust-lang/rust-clippy/issues/6446
-    #[allow(clippy::await_holding_lock)]
     pub(crate) async fn get_or_fetch(
         &self,
-        k: &K,
+        key: &K,
         fetch_fn: impl Fn() -> DbRecv<V>,
     ) -> DbResult<V> {
-        // See below comment about control flow.
-        let (mut slot_lock, complete_tx) = {
-            let mut cache = { self.cache.lock() };
-            if let Some(entry_lock) = cache.get(k).cloned() {
-                drop(cache);
-                let entry = entry_lock.read().await;
-                return entry.get_async().await;
-            }
-
-            // Create a new cache slot and insert and lock it.
-            let (complete_tx, complete_rx) = broadcast::channel(1);
-            let slot = Arc::new(RwLock::new(SlotState::Pending(complete_rx)));
-            cache.push(k.clone(), slot.clone());
-            let lock = slot
-                .try_write_owned()
-                .expect("cache: lock fresh cache entry");
+        if let Some(v) = self.get(key) {
+            return Ok(v);
+        }
 
-            (lock, complete_tx)
+        // Get the mutex
+        let mutex = {
+            let mut fetch_guard = self.fetch_mutexes.write();
+            fetch_guard
+                .entry(key.clone())
+                .or_insert_with(|| Arc::new(Mutex::new(())))
+                .clone()
         };
+        // If not, need to fetch
+        let _guard = mutex.lock().await;
 
-        let res = match fetch_fn().await {
-            Ok(Ok(v)) => v,
-            Ok(Err(e)) => {
-                error!(?e, "failed to make database fetch");
-                *slot_lock = SlotState::Error;
-                self.purge(k);
-                return Err(e);
+        // After acquiring the lock, check for the value
+        if let Some(value) = self.get(key) {
+            // Value present, no need to call fetch_fn
+            return Ok(value);
+        }
```

Comment:
This requires taking a lock on the LRU cache again, doubling the lock contention, which is not ideal.